Add _page to ACPAPI Directory resource proposal

tdahlgre opened the pull request:
@mburbidgadobe, @trieloff, @lopperma, others - This WIP change proposal adds support for _page for Directory resource per ACPAPI specification.

See inline notes for additional details, and Platform API specification for pagination support (I'm not posting links or copies of internal references to this pull in public git location).

npm test passes locally with these changes, includes new directory sample and valid/invalid json examples for newly added _page object schema.

First time working in XDM so any feedback, guidelines and proposals welcome to make this good!

Questions:
- should there be git issue created for this (there is internal backlog item in Adobe backlog)
- should there be copy of the same changes in docs/ directories (I see some, mostly duplicate, or is it somehow out-generated)?
-- Tuomas: got answer to this, no need to, it will be automatically generated.

tdahlgre commented on schemas/external/repo/directory-page.example.1.json:
```diff
@@ -0,0 +1,7 @@
+{
```
sample with minimal required properties

tdahlgre commented on schemas/external/repo/directory-page.example.2.json:
```diff
@@ -0,0 +1,8 @@
+{
```
sample with optional properties included

tdahlgre commented on schemas/external/repo/directory-page.invalid.1.json:
```diff
@@ -0,0 +1,6 @@
+{
```
invalid example, required 'count' missing.

tdahlgre commented on schemas/external/repo/directory-page.schema.json:
```diff
@@ -0,0 +1,67 @@
+{
```
page-object as defined in ACPAPI for directories. This is now separate schema file located in external/repo folder, could also be defined in separate folder like 'hal' links are or moved to some other 'common' location as well ... feedback and proposals welcome.

fmeschbe commented on schemas/external/repo/directory-page.schema.json:
```diff
@@ -0,0 +1,67 @@
+  ],
+  "required": [
+    "orderBy",
+    "order",
```
The order property is optional and actually is not officially listed in the [Sorting, Paging, and Filtering](https://git.corp.adobe.com/pages/AdobeCloudPlatform/api-spec/#sorting-paging-and-filtering) Spec any more. Rather its functionality has been replaced by the `orderby`  property supporting multiple values with optional `+/-` to indicate sort order.

Not sure, maybe we should remove this property alltogether.

tdahlgre replied:
@fmeschbe I do see 'order' listed in current spec, and I assumed that all the properties that do not have 'if any' are mandatory. I can removed or make this optional, and matching changes should be applied to spec as well.

Here's copy of current spec and page-object definition;
![image](https://user-images.githubusercontent.com/35572085/35286461-ee114d6e-0014-11e8-949f-9e0685eb6db3.png)

fmeschbe replied:
Thanks. I suggest to make at least optional for now.

fmeschbe commented on schemas/external/repo/directory-page.schema.json:
```diff
@@ -0,0 +1,67 @@
+  "definitions": {
+    "page": {
+      "properties": {
+        "orderBy": {
```
The API spec has this all lowercase as `orderby`.  As we have naming convention to be lowerCamelCase for consistency `orderBy` would be correct.
@ogoldman, @bpruessm what do you think ?

trieloff commented:
The `docs` will get auto-generated once it hits master. You don't need to include it in your PR, but you can run `npm run markdown` to see what it would look like.

fmeschbe commented on schemas/external/repo/directory-page.schema.json:
```diff
@@ -0,0 +1,67 @@
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Directory Page Object",
+  "type": "object",
+  "description": "Directory page object contains paging information for related directory when directory result is paginated.",
```
I think we should not called this with any relation to *Directory* as it is also used vor the `versions` resource and IMHO is a generic model for paging collection-style objects.

tdahlgre replied:
Will move page as common definition.

fmeschbe commented:
As I suggest this to be a generic model, we might even consider moving it to the `schemas/common` folder, for example ?

fmeschbe commented:
@trieloff I suggested to move this schema to `schemas/common` as I think it is not restricted to `extensions/repo` only.

Please advise whether this is correct/ok or what would be better ?

Thanks.

tdahlgre commented on schemas/external/page/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+      "$ref": "#/definitions/page"
+    }
+  ],
+  "required": [
```
order now updated to be optional per recommendation from @fmeschbe.

tdahlgre commented on schemas/external/page/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+{
```
@trieloff - this schema is now in schemas/external/page folder (following standard seen with schemas/external/hal to have folder per common type. @fmeschbe proposed moving this to schemas/commons folder to be even more generic. Please advise if that would be desired location and I'll move it there.

trieloff replied:
I think common sounds right. It's not really an external standard (like HAL that we re-implemented), but very common.

trieloff commented:
@fmeschbe please take another look. 

I won't be able to merge earlier than Friday, because I'm at a conference tomorrow.

tdahlgre commented on schemas/common/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/xdm/common/page",
```
@fmeschbe, @trieloff now page definition moved to schemas/commons folder and id reference updated to 'xdm/common' like used in other definitions in schemas/commons.

tdahlgre commented on schemas/external/repo/directory.example.2.json:
```diff
@@ -0,0 +1,25 @@
+      "templated": true
+    }
+  },
+  "_page": {
```
I also added another directory sample which includes _page

trieloff commented on schemas/common/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+        "order": {
+          "title": "Sort order",
+          "type": "string",
+          "pattern": "^(asc|desc)$",
```
Better: `"enum": [ "asc", "desc" ]` and 
`"meta:enum": {"asc": "Ascending order, i.e. the entries with the greatest value according to `orderBy` come first.", "desc": "Descending order"}`

fmeschbe replied:
+1 for `meta:enum`, though this property is likely going away.

trieloff commented on schemas/common/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+        "start": {
+          "title": "First value",
+          "type": "string",
+          "description": "The first value, in sorted order, of the orderby property on this page.\n"
```
Make sure `orderBy` is in camelcase, even in descriptions. Use backticks to make it stand out. (the `\n` at the end is not needed)

trieloff replied:
Question: is `orderBy` only allowed for `string` values? If no, the type of this property might be `anyOf` string, number, boolean (at least)

fmeschbe replied:
`orderBy` is a comma-separated list of property names, optionally prefixed with `+` or `-`. Hence it is a `string` always.

trieloff replied:
`orderBy` is always a `string`, that much is clear. The question is if ordering is only allowed for properties that have string values (which does not seem to make sense). In that case, the `start` and `next` properties should be able to have other values, too.

trieloff replied:
What does `+` and `-` mean? All of that needs to go into the description string.

trieloff commented on schemas/common/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+        "property": {
+          "title": "Filter properties",
+          "type": "string",
+          "description": "The list of properties by which the result is filtered, if any.\n"
```
If it's a list, what is the list separator? What is the escape sequence in case the list separator is used in a property name?

In cases like this, it's a good idea to also include an `examples` array right in the property.

fmeschbe replied:
Comma-separated list if property names („tokens“). Good question what the character space for „token“ actually is. I would suspect comma is not a valid token character.

trieloff replied:
If I understood you correct, then `property` simply checks for existence of the property. What counts as non-existence?

- `undefined`
- `null`
- `0`
- `false`
- `NaN`

trieloff commented on schemas/common/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+        "type": {
+          "title": "Filter type value",
+          "type": "string",
+          "description": "The applied type filtering value, if any.\n"
```
How is `type` related to `filter`? Is this just a special, inconsistent way to filter for `dc:type` values?

fmeschbe replied:
`type` filters on the media type of the list entry, while `property` filters on the existence of a specific property in a list entry.

fmeschbe commented on schemas/common/page.schema.json:
```diff
@@ -0,0 +1,67 @@
+          "type": "string",
+          "description": "The property by which this page is ordered.\n"
+        },
+        "order": {
```
FYI: this property has now been removed and we should probably also removed it here from the _page object. It serves not useful purpose any more. Thanks.

/cc @tdahlgre, @trieloff

trieloff commented:
@tdahlgre, this PR still has unaddressed issues and change requests from @fmeschbe. Is there anything I can do to help move it forward?

tdahlgre commented:
@trieloff, @fmeschbe - I have removed 'order' property and added description to orderBy per description in Platform API spec. Let me know if there is other items to be addressed. I also removed "WIP" from pull request title

trieloff commented:
Looks good from my end. I've updated the documentation and filled in the missing bits.

@fmeschbe care to remove the red x?

trieloff commented:
@fmeschbe unless I hear from you by tomorrow, I'll get this ready to be merged.
